File: parser.py
from urllib import request
import re

site = 'http://www.mosigra.ru'
#site = 'http://www.csd.tsu.ru'
ITERATIONS = 10

mailPattern = '[\w]+@[\w.]+[\w]'
urlPatternt = '(?:' + site + ')?' + '/(?:[a-zA-Z]|[0-9]|[\/\.\?\-])+'
hrefPatternt = '(?<=a href=")'

maillist = []
start = [site]
urllist = [site]
index = 0

def req(urls):
    global index
    urllist.extend(urls)
    #print(urls)
    for i in urls:
        if index < ITERATIONS - 1:
            try:
                if i[:1] == '/':
                    html = request.urlopen(site + i).read().decode('utf-8')
                else:
                    html = request.urlopen(i).read().decode('utf-8')
            except request.HTTPError as e:
                #print(e.code, e.msg)
                continue
            except UnicodeDecodeError:
                #print("Oops, file")
                continue
            index += 1

            newmails = re.findall(mailPattern, html)
            maillist.extend(newmails)

            applicant = re.findall(hrefPatternt + urlPatternt, html)
            urlsNew = set(applicant) - set(urllist)
            req(list(urlsNew))

req(start)
print(set(maillist))

File: parser.rb
require 'net/http'
require 'uri'
require 'nokogiri'

$site = "http://www.mosigra.ru"
$Iterations = 10
$mailPattern = Regexp.new('[\w]+@[\w.]+[\w]')
start = [$site]
$urllist = [$site]
$index = 0
$mailslist = []
def req(urls)
  for url in urls do
    if $index < $Iterations - 1
      if url[0] == '/'
        url = $site+url
      end
      $index+=1
      uri = URI.parse(url)
      response = Net::HTTP.get_response(uri)
      content = response.body
      doc = Nokogiri::HTML(content)
      emails = content.scan($mailPattern).uniq
      $mailslist.push(emails)
      list  = doc.css("a")
      ur = Array.new
      list.each do |index|
        if index['href'].to_s.scan(/[@$#\s]/).size == 0 and index['href'].to_s.size > 1
          ur.push(index['href'].to_s)
        end
      end
      ur=ur-$urllist
      $urllist << ur
      req(ur)
    else
      break
    end
  end
end

req(start)
puts($mailslist.uniq)
